presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -675,14 +679,18 @@ public Optional<Partition> getPartition(MetastoreContext metastoreContext, Strin
     }
 
     @Override
-    public Optional<List<String>> getPartitionNames(MetastoreContext metastoreContext, String databaseName, String tableName)
+    public Optional<List<String>> getPartitionNames(MetastoreContext metastoreContext, Table table)
     {
-        return get(partitionNamesCache, getCachingKey(metastoreContext, hiveTableName(databaseName, tableName)));
+        return get(partitionNamesCache, getCachingKey(metastoreContext, hiveTableName(table.getDatabaseName(), table.getTableName())));
     }
 
     private Optional<List<String>> loadPartitionNames(KeyAndContext<HiveTableName> hiveTableNameKey)
     {
-        return delegate.getPartitionNames(hiveTableNameKey.getContext(), hiveTableNameKey.getKey().getDatabaseName(), hiveTableNameKey.getKey().getTableName());
+        Optional<Table> table = getTable(hiveTableNameKey.getContext(), hiveTableNameKey.getKey().getDatabaseName(), hiveTableNameKey.getKey().getTableName());
+        if (!table.isPresent()) {
```

Comment:
We should try to move this check to the respective metastore implementation since Thrift and Glue both implementations behave differently in the current scenario. ThriftHiveMetastore doesn't throw an exception whereas GlueHiveMetastore does.

Reply:
This requires making change in getPartitionNames method to pass Optional<Table> instead of Table as an argument. But this is a valid ask and hence I will change the method signature.